### PR TITLE
Add soft-landing penalization as per Olaf paper

### PR DESCRIPTION
**Pull request by saikishor**
This implements the soft landing explains in the Olaf paper : https://arxiv.org/pdf/2512.16705

### CODE REVIEW

File: src/pal_mjlab/tasks/velocity/mdp/rewards.py [hunk 1]
```diff
@@ -417,5 +418,96 @@ def __call__(
...
     env.extras["log"]["Metrics/joint_vel_limit_violation"] = torch.mean(penalty).item()
-    
-    return penalty
\ No newline at end of file
+
+    return penalty
+
+
+class disney_soft_landing:
+  """Penalize high impact forces at landing to encourage soft footfalls similar to Disney Olaf."""
```
Comment (louislelay) [inline src/pal_mjlab/tasks/velocity/mdp/rewards.py]:
```suggestion
class sound_suppression:
  """Soft-landing reward that penalizes impulsive contact forces at footstrike.
  
  Ref: https://arxiv.org/abs/2512.16705 (Disney Olaf paper)
  """
```

File: src/pal_mjlab/tasks/velocity/mdp/rewards.py [hunk 2]
```diff
@@ -417,5 +418,96 @@ def __call__(
...
+    max_component_z = squared_change_in_site_velocities[..., 2]
+
+    max_value = torch.max(torch.max(max_component_x, max_component_y), max_component_z)
+    max_of_all_sites = torch.max(max_value, dim=1).values  # shape (N,)
+    # print(f"Max value x: {max_component_x}, y: {max_component_y}, z: {max_component_z}, max_value:"
+    # " {max_value} and max_of_all_sites : {max_of_all_sites}")
+
+    change_in_velocities_along_projected_gravity = squared_term_along_proj_gravity[
+      ..., 2
+    ]  # shape (N, M)
```
Comment (louislelay) [inline src/pal_mjlab/tasks/velocity/mdp/rewards.py:475]: > "An impact-reduction term complements the reward function and penalizes changes in velocity along the gravity direction between simulation steps, thereby producing
smoother foot motions and reducing impact noise." -- Paragraph 8 from Section 5.C (https://arxiv.org/abs/2512.16705)

The paper states the reward "penalizes changes in velocity along the gravity direction," and the formula uses $\Delta v_{i,z}$. I might be wrong but wouldn't that mean that $z$ is simply the world-frame z-axis?

Since we're already reading `site_lin_vel_w` (world frame), the z-component of the velocity difference directly gives us $\Delta v_{i,z}$, no projection onto body-frame gravity would then be needed. IMO if a dot product with projected gravity were intended, the paper would have formulated it as such rather than using a plain $z$ subscript.

That simplifies this part to:

```python
delta_v_z = change_in_site_velocities[..., 2]
```

File: src/pal_mjlab/tasks/velocity/mdp/rewards.py [hunk 3]
```diff
@@ -417,5 +418,96 @@ def __call__(
...
+        max_of_all_sites.unsqueeze(1),
+        change_in_velocities_along_projected_gravity,
+      ),
+      dim=1,
+    )
+
+    # print(f"The z_component : {change_in_velocities_along_projected_gravity} and max_component : {max_of_all_sites} and "
+    # "cost : {cost} and the term is : {term} and squared_term_along_proj_gravity : {squared_term_along_proj_gravity}")
+
+    # return the squared sum of change in velocity along the projected gravity direction
```
Comment (louislelay) [inline src/pal_mjlab/tasks/velocity/mdp/rewards.py]: > "Because the physics engine can generate large velocity changes during contact resolution, which may result in a large reward term and destabilize critic learning, we saturate the impact reduction reward term in the reward computation." -- Paragraph 8 from Section 5.C (https://arxiv.org/abs/2512.16705)

The paper introduces a fixed $\Delta v_{\max}^2$ to clamp the per-foot penalty: $\min(\Delta v_{i,z}^2, \Delta v_{\max}^2)$. The purpose is to filter out unrealistically large velocity spikes that the physics engine (Isaac Sim) produces during contact resolution.

The current implementation computes `max_of_all_sites` as the maximum squared velocity change across all axes and sites at the current step, then takes `min(max_of_all_sites, delta_v_z²)`. Since `max_of_all_sites` is by definition ≥ any individual component, this min is effectively a no-op -- the crazy spikes from contact resolution pass through unclamped, which is exactly what the paper is trying to prevent.

Instead, $\Delta v_{\max}$ should be a fixed hyperparameter. The paper doesn't report its value, but we can reason about a sensible range. It should be large enough to not interfere with normal foot strikes, but small enough to reject physics engine artifacts. If you agree with this, let's proceed with finding this reasonable value.

**Review by louislelay — CHANGES_REQUESTED** [review]
Hello @saikishor, I was reading through the paper alongside your implementation and I differ on the understanding of it. Tell me WDYT!

File: src/pal_mjlab/tasks/velocity/mdp/rewards.py (hunk 3 above)
Comment (saikishor) [reply src/pal_mjlab/tasks/velocity/mdp/rewards.py]: I disagree, delta v_max cannot be a hyperparameter. Atleast, IMO

File: src/pal_mjlab/tasks/velocity/mdp/rewards.py (hunk 2 above)
Comment (saikishor) [reply src/pal_mjlab/tasks/velocity/mdp/rewards.py:475]: Understood, it could be that it works only for flat terrain?. I'll train one more with gravity vector and see how it will work on Monday

**Review by louislelay — APPROVED** [review]
